### PR TITLE
Speed up the initialization of ApplicationEngine.

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

File: src/neo/SmartContract/ApplicationEngine.cs [hunk 1]
```diff
@@ -26,13 +26,11 @@ private class InvocationState
         public static event EventHandler<NotifyEventArgs> Notify;
         public static event EventHandler<LogEventArgs> Log;
 
-        public const long GasFree = 0;
```
Comment (shargon) [inline src/neo/SmartContract/ApplicationEngine.cs:29]: We will never have a free tx?

File: src/neo/SmartContract/ApplicationEngine.cs [hunk 2]
```diff
@@ -45,15 +43,21 @@ private class InvocationState
...
+        public IReadOnlyList<NotifyEventArgs> Notifications => notifications ?? (IReadOnlyList<NotifyEventArgs>)Array.Empty<NotifyEventArgs>();
 
         public ApplicationEngine(TriggerType trigger, IVerifiable container, StoreView snapshot, long gas, bool testMode = false)
+            : this(trigger, container, snapshot, gas, new ReferenceCounter())
         {
-            this.gas_amount = GasFree + gas;
             this.testMode = testMode;
+        }
+
+        private ApplicationEngine(TriggerType trigger, IVerifiable container, StoreView snapshot, long gas, ReferenceCounter referenceCounter)
```
Comment (shargon) [inline src/neo/SmartContract/ApplicationEngine.cs:54]: We can't share a referenceCounter under test mode? the wallet will need to replicate the same behaviour.

File: src/neo/SmartContract/ApplicationEngine.cs (hunk 1 above)
Comment (erikzhang) [reply src/neo/SmartContract/ApplicationEngine.cs:29]: If we have free transactions, then the value of `GasFree` for `Application` and `Verification` is also different. This field is not suitable for being included in `ApplicationEngine`.

**Comment by erikzhang** [comment]
I will close it because we can use `PostPersist()` to solve the problem of oracle.

f8d3362

**Comment by erikzhang** [comment]
I will remove the constructor for ReferenceCounter sharing and keep the rest.

**Pull request merged**